[TIMOB-23569] Android: Implement method to handle webview input type=file

ashcoding opened the pull request:
https://jira.appcelerator.org/browse/TIMOB-23569

Commit:
[TIMOB-19463] Android: Implement method to handle webview input type=file

ashcoding committed:
[TIMOB-23229] Required changes for compiling

hieupham007 commented on android/modules/ui/src/java/ti/modules/titanium/ui/widget/webview/TiWebChromeClient.java:
```diff
@@ -172,5 +188,236 @@ public boolean interceptOnBackPressed()
+	    mFilePathCallbackLegacy = filePathCallback;
+
+	    Intent takePictureIntent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
+	    if (takePictureIntent.resolveActivity(tiWebView.getProxy().getActivity().getPackageManager()) != null) {
```
hmm, this'll crash if the proxy or activity is null.

hieupham007 commented on android/modules/ui/src/java/ti/modules/titanium/ui/widget/webview/TiWebChromeClient.java:
```diff
@@ -172,5 +188,236 @@ public boolean interceptOnBackPressed()
+	    Intent chooserIntent = new Intent(Intent.ACTION_CHOOSER);
+	    chooserIntent.putExtra(Intent.EXTRA_INTENT, contentSelectionIntent);
+	    chooserIntent.putExtra(Intent.EXTRA_TITLE, "Image Chooser");
+	    chooserIntent.putExtra(Intent.EXTRA_INITIAL_INTENTS, intentArray);
```
this could be an empty array. I don't think we should be adding empty objects to the intent, unless it serves some purpose.

hieupham007 commented on android/modules/ui/src/java/ti/modules/titanium/ui/widget/webview/TiWebChromeClient.java:
```diff
@@ -172,5 +188,236 @@ public boolean interceptOnBackPressed()
+        mFilePathCallback = filePathCallback;
+
+        Intent takePictureIntent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
+        if (takePictureIntent.resolveActivity(tiWebView.getProxy().getActivity().getPackageManager()) != null) {
```
could crash on edge cases, see above comment.

hieupham007 commented on android/modules/ui/src/java/ti/modules/titanium/ui/widget/webview/TiWebChromeClient.java:
```diff
@@ -172,5 +188,236 @@ public boolean interceptOnBackPressed()
+            // Create the File where the photo should go
+            File photoFile = null;
+            try {
+                photoFile = createImageFile();
```
There's quite a bit of duplicate code for onShowFileChooser and openFileChooser. Could we implement a method that has some shared common code?

ashcoding committed:
[TIMOB-23229] Addressing comments

ashcoding committed:
[TIMOB-23229] Implement common code for onShowFileChooser and openFileChooser to use

ashcoding commented:
@hieupham007 comments addressed.

ashcoding commented on android/modules/ui/src/java/ti/modules/titanium/ui/widget/webview/TiWebChromeClient.java:
```diff
@@ -172,5 +190,234 @@ public boolean interceptOnBackPressed()
+        }
+    }
+
+    private File createImageFile() throws IOException {
```
On an Android 6.0+, you get the error,
```
[ERROR] : TiWebChromeClient: (main) [9148,23167] Unable to create Image File
[ERROR] : TiWebChromeClient: java.io.IOException: open failed: EACCES (Permission denied)
```
This should be handled.

ashcoding replied:
New push has handled this issue.

ashcoding commented:
As mentioned in the comments, we need to handle that by checking if the permission is granted at runtime. If it is not, do not add Camera as one of the choices.

ashcoding committed:
[TIMOB-23569] Android: Image Capture

The Image Capture intent is only included if you are using Android API
22 and below OR if 'Write External Storage' permission is granted while
using API 23 and above.

If API 23 and above is used and no permission is granted, the intent
will be null and there will be no additional option for using the
Camera.

ashcoding commented:
Comments addressed.

hieupham007 commented:
Code reviewed and functionally tested. Request **accepted**
